text/0000-buildpack-uris.md

```diff
@@ -0,0 +1,189 @@
+- `urn:cnb:registry:bp.id@bp.version` - A reference to a buildpack with id `bp.id` at the specific version `bp.version`.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+### pack
+
+Certain `pack` CLI arguments would take URIs supported instead of the current inconsistent format.
+
+Some examples:
+
+```shell script
+pack build my-app \
+  --buildpack urn:cnb:builder:bp.id@bp.version \                  # buildpack from builder
+  --buildpack file:///home/user/path/to/buildpack/ \              # absolute via file
+  --buildpack /home/user/path/to/buildpack/ \                     # absolute via schemeless
+  --buildpack ../path/to/buildpack/ \                             # relative file
+  --buildpack docker:/cnbs/some-package \                         # Docker Hub image
+  --buildpack docker://gcr.io/cnbs/sample-package-2:bionic \      # GCR image (with tag)
+  --buildpack cnb://index.buildpack.io/my-org/bp.id@bp.version \  # buildpack from registry
+  --buildpack urn:cnb:builder                                     # All buildpacks in builder
+```   
+
+### Config Files
+
+Config files would replace `image` with `uri`, using the `docker://` scheme, where applicable.
+
+`package.toml`:
+```toml
+[buildpack]
+uri = "."
+
+[[dependencies]]
+uri = "docker:/my/image:latest"
+
+# ...
+``` 
+
+`builder.toml`:
+```toml
+# ...
+
+[[buildpacks]]
+uri = "./my/buildpacks/"
+
+[[buildpacks]]
+uri = "docker:/my/image:latest"
+
+# ...
+```
+
+# Additional considerations
+
+#### `pack` User Experience
+
+`pack` could provide additional logic in order for users to have a nicer user experience when providing references
+via CLI. For example, `pack` could have an order of precedence to make a best guess at which type of reference
+they are referring to or provide additional feedback if it's ambiguous.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The user would need to type an additional scheme prefix for any URI which is not a relative reference.
+
+# Alternatives
+[alternatives]: #alternatives
+
+#### Flags and Config Keys
+
+A different flag or config properties to signify the different protocols and would render the need for
+URI moot.
+
+#### Docker scheme
+
+An alternative to `docker` is to use a more generic term such as `oci`. There is an official
+[OCI distribution spec](https://github.com/opencontainers/distribution-spec/blob/master/spec.md) based on Docker HTTP
+API v2 but it's direct compatibility would require additional research. Additionally, `oci` scheme has an overlap with
+[Oracle Cloud Infrastructure Object Storage Service](https://docs.cloud.oracle.com/en-us/iaas/tools/hdfs/2.9.2.1/). 
+
+# Prior Art
+[prior-art]: #prior-art
+
+- Predefined set of URI schemes for OCI images created by [Skopeo](https://github.com/containers/skopeo).
+- [from=builder][from-builder] syntax.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Where would this be documented long-term?
```

Comment:
I wonder if we should move this to an extension spec or part of the [project.toml one](https://github.com/buildpacks/spec/pull/70).

Reply:
Update to detail extension spec.